src/librustc_mir/transform/generator.rs

```diff
@@ -539,9 +539,23 @@ fn locals_live_across_suspend_points(
             let mut live_locals_here = storage_required;
             live_locals_here.intersect(&liveness.outs[block]);
 
-            // The generator argument is ignored
+            // The generator argument is ignored.
             live_locals_here.remove(self_arg());
 
+            // If the resume arg ends up in a plain local, we ignore the local's liveness in the
+            // code before the `Yield`. This makes it so the local doesn't end up in the generator
+            // state, which isn't needed as it will be assigned a new value during the next
+            // resumption of the generator.
+            // This only works for `Yield`s that store into a plain local without projections
+            // though: If it would store to one element of a tuple, for example, the other half
+            // might need to be stored in the state.
+            // This is a slight hack that should probably be removed once we have more precise
+            // analyses. In particular the dataflow/liveness analyses don't know that `Yield` only
+            // stores to the `resume_arg` place upon resumption (and never reads from it).
+            if resume_arg.projection.is_empty() {
+                live_locals_here.remove(resume_arg.local);
```

Comment:
Should this case on whether or not the local being written to by the resume argument implements `Drop` and has not been invalidated? That is, `let mut x = foo.lock(); x = yield;` shouldn't release the lock until after the `yield` has returned a value, so we need to continue to store space for `x` in the generator state, right?

Reply:
I believe this is all handled by MIR construction / drop elaboration already, at this point MIR-level assignments just overwrite the old value without running any drop glue. Specifically in your example, the resume argument would be assigned to a temporary first, which then overwrites `x` after the old value has been dropped.

Reply:
I'd rather not rely on more implementation details of MIR construction, if possible.

Reply:
It's not really an implementation detail – assignments (and terminators that write to a place) simply don't drop the old value

Reply:
Sorry, after re-reading I think @cramertj's case was slightly different than the one I had in mind (the one I commented on below).